File: 2019-01-03/rangeSumBST.cc
int rangeSumBST(TreeNode* root, int L, int R) {
    int sum = 0;

    if (root == NULL) return sum;

    if (root->val < L && root->right) {
        return sum += rangeSumBST(root->right,L,R);
    }
    if (R < root->val && root->left) {
        return sum += rangeSumBST(root->left,L,R);
    }
    if (root->val >= L && root->val <= R) {
        // For leetcode's test set the conditionals are faster than combining the three +='s
        sum += root->val;

        if (root->left) {
            sum += rangeSumBST(root->left,L,R);
        }
        if (root->right) {
            sum += rangeSumBST(root->right,L,R);
        }
    }

  return sum;
}

File: 2019-01-03/rangeSumBST.js
function rangeSum(node, sum, L, R) {
    if (!node) return 0;
    if (node.val >= L && node.val <= R) {
        sum += node.val + rangeSum(node.left, sum, L, R) + rangeSum(node.right, sum, L, R);
        return sum;
    }
    if (node.val < L) return rangeSum(node.right, sum, L, R);
    if (node.val > R) return rangeSum(node.left, sum, L, R);
    return 0;
}

var rangeSumBST = function(root, L, R) {
    var sum = 0;
    sum = rangeSum(root, sum, L, R);
    return sum;
};